// @splitsoftware/splitio/lib/sync/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ControlTypes = exports.OCCUPANCY = exports.CONTROL = exports.SPLIT_UPDATE = exports.SPLIT_KILL = exports.SEGMENT_UPDATE = exports.MY_SEGMENTS_UPDATE = exports.SSE_ERROR = exports.PUSH_DISCONNECT = exports.PUSH_DISABLED = exports.PUSH_CONNECT = exports.SECONDS_BEFORE_EXPIRATION = void 0;
// time for refresh token
var SECONDS_BEFORE_EXPIRATION = 600; // Internal SDK events, subscribed by SyncManager and PushManager

exports.SECONDS_BEFORE_EXPIRATION = SECONDS_BEFORE_EXPIRATION;
var PUSH_CONNECT = 'PUSH_CONNECT';
exports.PUSH_CONNECT = PUSH_CONNECT;
var PUSH_DISABLED = 'PUSH_DISABLED';
exports.PUSH_DISABLED = PUSH_DISABLED;
var PUSH_DISCONNECT = 'PUSH_DISCONNECT';
exports.PUSH_DISCONNECT = PUSH_DISCONNECT;
var SSE_ERROR = 'SSE_ERROR'; // Update-type push notifications, handled by NotificationProcessor

exports.SSE_ERROR = SSE_ERROR;
var MY_SEGMENTS_UPDATE = 'MY_SEGMENTS_UPDATE';
exports.MY_SEGMENTS_UPDATE = MY_SEGMENTS_UPDATE;
var SEGMENT_UPDATE = 'SEGMENT_UPDATE';
exports.SEGMENT_UPDATE = SEGMENT_UPDATE;
var SPLIT_KILL = 'SPLIT_KILL';
exports.SPLIT_KILL = SPLIT_KILL;
var SPLIT_UPDATE = 'SPLIT_UPDATE'; // Control-type push notifications, handled by NotificationKeeper

exports.SPLIT_UPDATE = SPLIT_UPDATE;
var CONTROL = 'CONTROL';
exports.CONTROL = CONTROL;
var OCCUPANCY = 'OCCUPANCY';
exports.OCCUPANCY = OCCUPANCY;
var ControlTypes = {
  STREAMING_DISABLED: 'STREAMING_DISABLED',
  STREAMING_PAUSED: 'STREAMING_PAUSED',
  STREAMING_RESUMED: 'STREAMING_RESUMED'
};
exports.ControlTypes = ControlTypes;
