// @splitsoftware/splitio-commons/cjs/storages/KeyBuilderSS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyBuilderSS = void 0;
var tslib_1 = require("tslib");
var KeyBuilder_1 = require("./KeyBuilder");
var methodNames = {
    t: 'treatment',
    ts: 'treatments',
    tc: 'treatmentWithConfig',
    tcs: 'treatmentsWithConfig',
    tr: 'track'
};
var KeyBuilderSS = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(KeyBuilderSS, _super);
    function KeyBuilderSS(prefix, metadata) {
        var _this = _super.call(this, prefix) || this;
        _this.metadata = metadata;
        return _this;
    }
    KeyBuilderSS.prototype.buildRegisteredSegmentsKey = function () {
        return this.prefix + ".segments.registered";
    };
    KeyBuilderSS.prototype.buildImpressionsKey = function () {
        return this.prefix + ".impressions";
    };
    KeyBuilderSS.prototype.buildEventsKey = function () {
        return this.prefix + ".events";
    };
    KeyBuilderSS.prototype.searchPatternForSplitKeys = function () {
        return this.buildSplitKeyPrefix() + "*";
    };
    /* Telemetry keys */
    KeyBuilderSS.prototype.buildLatencyKey = function (method, bucket) {
        return this.prefix + ".telemetry.latencies::" + this.buildVersionablePrefix() + "/" + methodNames[method] + "/" + bucket;
    };
    KeyBuilderSS.prototype.buildExceptionKey = function (method) {
        return this.prefix + ".telemetry.exceptions::" + this.buildVersionablePrefix() + "/" + methodNames[method];
    };
    KeyBuilderSS.prototype.buildInitKey = function () {
        return this.prefix + ".telemetry.init::" + this.buildVersionablePrefix();
    };
    KeyBuilderSS.prototype.buildVersionablePrefix = function () {
        return this.metadata.s + "/" + this.metadata.n + "/" + this.metadata.i;
    };
    return KeyBuilderSS;
}(KeyBuilder_1.KeyBuilder));
exports.KeyBuilderSS = KeyBuilderSS;
