// @splitsoftware/splitio-commons/cjs/storages/pluggable/TelemetryCachePluggable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetryCachePluggable = void 0;
var findLatencyIndex_1 = require("../findLatencyIndex");
var TelemetryCachePluggable = /** @class */ (function () {
    /**
     * Create a Telemetry cache that uses a storage wrapper.
     * @param log  Logger instance.
     * @param keys  Key builder.
     * @param wrapper  Adapted wrapper storage.
     */
    function TelemetryCachePluggable(log, keys, wrapper) {
        this.log = log;
        this.keys = keys;
        this.wrapper = wrapper;
    }
    TelemetryCachePluggable.prototype.recordLatency = function (method, latencyMs) {
        return this.wrapper.incr(this.keys.buildLatencyKey(method, (0, findLatencyIndex_1.findLatencyIndex)(latencyMs)))
            .catch(function () { });
    };
    TelemetryCachePluggable.prototype.recordException = function (method) {
        return this.wrapper.incr(this.keys.buildExceptionKey(method))
            .catch(function () { });
    };
    return TelemetryCachePluggable;
}());
exports.TelemetryCachePluggable = TelemetryCachePluggable;
