// @splitsoftware/splitio/lib/__tests__/nodeSuites/push-fallbacking.spec.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testFallbacking = testFallbacking;

var _splitchangesRealWithSegments = _interopRequireDefault(require("../mocks/splitchanges.real.withSegments.json"));

var _splitchangesRealUpdateWithSegments = _interopRequireDefault(require("../mocks/splitchanges.real.updateWithSegments.json"));

var _splitchangesRealUpdateWithoutSegments = _interopRequireDefault(require("../mocks/splitchanges.real.updateWithoutSegments.json"));

var _messageOCCUPANCY0Control_pri = _interopRequireDefault(require("../mocks/message.OCCUPANCY.0.control_pri.1586987434550.json"));

var _messageOCCUPANCY1Control_pri = _interopRequireDefault(require("../mocks/message.OCCUPANCY.1.control_pri.1586987434450.json"));

var _messageOCCUPANCY1Control_sec = _interopRequireDefault(require("../mocks/message.OCCUPANCY.1.control_sec.1586987434451.json"));

var _messageOCCUPANCY2Control_pri = _interopRequireDefault(require("../mocks/message.OCCUPANCY.2.control_pri.1586987434650.json"));

var _messageCONTROLSTREAMING_PAUSEDControl_pri = _interopRequireDefault(require("../mocks/message.CONTROL.STREAMING_PAUSED.control_pri.1586987434750.json"));

var _messageCONTROLSTREAMING_RESUMEDControl_pri = _interopRequireDefault(require("../mocks/message.CONTROL.STREAMING_RESUMED.control_pri.1586987434850.json"));

var _messageCONTROLSTREAMING_DISABLEDControl_pri = _interopRequireDefault(require("../mocks/message.CONTROL.STREAMING_DISABLED.control_pri.1586987434950.json"));

var _messageSPLIT_UPDATE = _interopRequireDefault(require("../mocks/message.SPLIT_UPDATE.1457552649999.json"));

var _messageSEGMENT_UPDATE = _interopRequireDefault(require("../mocks/message.SEGMENT_UPDATE.1457552640000.json"));

var _messageSEGMENT_UPDATE2 = _interopRequireDefault(require("../mocks/message.SEGMENT_UPDATE.1457552650000.json"));

var _authPushEnabledNode = _interopRequireDefault(require("../mocks/auth.pushEnabled.node.json"));

var _testUtils = require("../testUtils");

var _eventSourceMock = _interopRequireWildcard(require("../../sync/__tests__/mocks/eventSourceMock"));

var _node = require("../../services/getEventSource/node");

var _index = require("../../index");

var _settings = _interopRequireDefault(require("../../utils/settings"));

/**
 * Validate the handling of OCCUPANCY and CONTROL events
 */
// since: -1, till: 1457552620999 (for initial fetch)
// since: 1457552620999, till: 1457552649999 (for SPLIT_UPDATE event)
// since: 1457552649999, till: 1457552669999 (for second polling fetch)
var key = 'nicolas@split.io';
var baseUrls = {
  sdk: 'https://sdk.push-synchronization/api',
  events: 'https://events.push-synchronization/api',
  auth: 'https://auth.push-synchronization/api'
};
var config = {
  core: {
    authorizationKey: '<fake-token-push-1>'
  },
  scheduler: {
    featuresRefreshRate: 0.2,
    segmentsRefreshRate: 0.25,
    metricsRefreshRate: 3000,
    impressionsRefreshRate: 3000
  },
  urls: baseUrls,
  streamingEnabled: true // debug: true,

};
var settings = (0, _settings.default)(config);
var MILLIS_SSE_OPEN = 100;
var MILLIS_STREAMING_DOWN_OCCUPANCY = MILLIS_SSE_OPEN + 100;
var MILLIS_SPLIT_UPDATE_EVENT_DURING_POLLING = MILLIS_STREAMING_DOWN_OCCUPANCY + 100;
var MILLIS_STREAMING_UP_OCCUPANCY = MILLIS_STREAMING_DOWN_OCCUPANCY + settings.scheduler.featuresRefreshRate + 100;
var MILLIS_SPLIT_UPDATE_EVENT_DURING_PUSH = MILLIS_STREAMING_UP_OCCUPANCY + 100;
var MILLIS_STREAMING_PAUSED_CONTROL = MILLIS_SPLIT_UPDATE_EVENT_DURING_PUSH + 100;
var MILLIS_SEGMENT_UPDATE_EVENT_DURING_POLLING = MILLIS_STREAMING_PAUSED_CONTROL + 100;
var MILLIS_STREAMING_RESUMED_CONTROL = MILLIS_STREAMING_PAUSED_CONTROL + settings.scheduler.featuresRefreshRate + 100;
var MILLIS_SEGMENT_UPDATE_EVENT_DURING_PUSH = MILLIS_STREAMING_RESUMED_CONTROL + 100;
var MILLIS_STREAMING_DISABLED_CONTROL = MILLIS_SEGMENT_UPDATE_EVENT_DURING_PUSH + 100;
var MILLIS_DESTROY = MILLIS_STREAMING_DISABLED_CONTROL + settings.scheduler.featuresRefreshRate * 2 + 100;
/**
 * Sequence of calls:
 *  0.0 secs: initial SyncAll (/splitChanges, /segmentChanges/*), auth, SSE connection
 *  0.1 secs: SSE connection opened -> syncAll (/splitChanges, /segmentChanges/*)
 *  0.2 secs: Streaming down (OCCUPANCY event) -> fetch due to fallback to polling (/splitChanges, /segmentChanges/*)
 *  0.3 secs: SPLIT_UPDATE event ignored
 *  0.4 secs: periodic fetch due to polling (/splitChanges)
 *  0.45 secs: periodic fetch due to polling (/segmentChanges/*)
 *  0.5 secs: Streaming up (OCCUPANCY event) -> syncAll (/splitChanges, /segmentChanges/*)
 *  0.6 secs: SPLIT_UPDATE event -> /splitChanges
 *  0.7 secs: Streaming down (CONTROL event) -> fetch due to fallback to polling (/splitChanges, /segmentChanges/*)
 *  0.8 secs: SEGMENT_UPDATE event ignored
 *  0.9 secs: periodic fetch due to polling (/splitChanges)
 *  0.95 secs: periodic fetch due to polling (/segmentChanges/*)
 *  1.0 secs: Streaming up (CONTROL event) -> syncAll (/splitChanges, /segmentChanges/*)
 *  1.1 secs: SEGMENT_UPDATE event -> /segmentChanges/*
 *  1.2 secs: Streaming down (CONTROL event) -> fetch due to fallback to polling (/splitChanges, /segmentChanges/*)
 *  1.4 secs: periodic fetch due to polling (/splitChanges)
 *  1.45 secs: periodic fetch due to polling (/segmentChanges/*)
 *  1.6 secs: periodic fetch due to polling (/splitChanges, /segmentChanges/*)
 *  1.7 secs: periodic fetch due to polling (/segmentChanges/*)
 *  1.7 secs: destroy client
 */

function testFallbacking(fetchMock, assert) {
  assert.plan(13);
  fetchMock.reset();
  (0, _node.__setEventSource)(_eventSourceMock.default);
  var start, splitio, client; // mock SSE open and message events

  (0, _eventSourceMock.setMockListener)(function (eventSourceInstance) {
    var expectedSSEurl = "".concat(settings.url('/sse'), "?channels=NzM2MDI5Mzc0_NDEzMjQ1MzA0Nw%3D%3D_segments,NzM2MDI5Mzc0_NDEzMjQ1MzA0Nw%3D%3D_splits,%5B%3Foccupancy%3Dmetrics.publishers%5Dcontrol_pri,%5B%3Foccupancy%3Dmetrics.publishers%5Dcontrol_sec&accessToken=").concat(_authPushEnabledNode.default.token, "&v=1.1&heartbeats=true");
    assert.equals(eventSourceInstance.url, expectedSSEurl, 'EventSource URL is the expected');
    setTimeout(function () {
      eventSourceInstance.emitOpen();
      eventSourceInstance.emitMessage(_messageOCCUPANCY1Control_pri.default);
      eventSourceInstance.emitMessage(_messageOCCUPANCY1Control_sec.default);
    }, MILLIS_SSE_OPEN); // open SSE connection after 0.1 seconds

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageOCCUPANCY0Control_pri.default);
    }, MILLIS_STREAMING_DOWN_OCCUPANCY); // send an OCCUPANCY event for switching to polling

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageSPLIT_UPDATE.default);
    }, MILLIS_SPLIT_UPDATE_EVENT_DURING_POLLING); // send a SPLIT_UPDATE event while polling, to check that we are ignoring it

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageOCCUPANCY2Control_pri.default);
    }, MILLIS_STREAMING_UP_OCCUPANCY); // send a OCCUPANCY event for switching to push

    setTimeout(function () {
      assert.equal(client.getTreatment(key, 'real_split'), 'on', 'evaluation of initial Split');
      client.once(client.Event.SDK_UPDATE, function () {
        assert.equal(client.getTreatment(key, 'real_split'), 'off', 'evaluation of updated Split');
      });
      eventSourceInstance.emitMessage(_messageSPLIT_UPDATE.default);
    }, MILLIS_SPLIT_UPDATE_EVENT_DURING_PUSH); // send a SPLIT_UPDATE event when push resumed, to check that we are handling it

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageCONTROLSTREAMING_PAUSEDControl_pri.default);
    }, MILLIS_STREAMING_PAUSED_CONTROL); // send a CONTROL event for switching to polling

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageSEGMENT_UPDATE.default);
    }, MILLIS_SEGMENT_UPDATE_EVENT_DURING_POLLING); // send a SEGMENT_UPDATE event while polling, to check that we are ignoring it

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageCONTROLSTREAMING_RESUMEDControl_pri.default);
    }, MILLIS_STREAMING_RESUMED_CONTROL); // send a CONTROL event for switching to push

    setTimeout(function () {
      assert.equal(client.getTreatment(key, 'real_split'), 'off', 'evaluation with initial segment');
      client.once(client.Event.SDK_UPDATE, function () {
        assert.equal(client.getTreatment(key, 'real_split'), 'on', 'evaluation with updated segment');
      });
      eventSourceInstance.emitMessage(_messageSEGMENT_UPDATE2.default);
    }, MILLIS_SEGMENT_UPDATE_EVENT_DURING_PUSH); // send a SEGMENT_UPDATE event when push resumed, to check that we are handling it

    setTimeout(function () {
      eventSourceInstance.emitMessage(_messageCONTROLSTREAMING_DISABLEDControl_pri.default);
      assert.equal(eventSourceInstance.readyState, _eventSourceMock.default.CLOSED, 'EventSource connection closed on STREAMING_DISABLED CONTROL event');
    }, MILLIS_STREAMING_DISABLED_CONTROL); // send a CONTROL event for disabling push and switching to polling

    setTimeout(function () {
      client.destroy().then(function () {
        assert.pass('client destroyed');
      });
    }, MILLIS_DESTROY); // destroy client after 0.6 seconds
  });
  fetchMock.getOnce(settings.url('/auth'), function (url, opts) {
    if (!opts.headers['Authorization']) assert.fail('`/auth` request must include `Authorization` header');
    assert.pass('auth success');
    return {
      status: 200,
      body: _authPushEnabledNode.default
    };
  }); // initial split and segment sync

  fetchMock.getOnce(settings.url('/splitChanges?since=-1'), {
    status: 200,
    body: _splitchangesRealWithSegments.default
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=-1'), {
    status: 200,
    body: {
      since: -1,
      till: 1457552620999,
      name: 'employees',
      added: [key],
      removed: []
    }
  }); // extra retry due to double request (greedy fetch until since === till)

  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552620999'), {
    status: 200,
    body: {
      since: 1457552620999,
      till: 1457552620999,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // split and segment sync after SSE opened

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), {
    status: 200,
    body: {
      splits: [],
      since: 1457552620999,
      till: 1457552620999
    }
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552620999'), {
    status: 200,
    body: {
      since: 1457552620999,
      till: 1457552620999,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // fetches due to first fallback to polling

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), {
    status: 200,
    body: {
      splits: [],
      since: 1457552620999,
      till: 1457552620999
    }
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552620999'), {
    status: 200,
    body: {
      since: 1457552620999,
      till: 1457552620999,
      name: 'employees',
      added: [],
      removed: []
    }
  });
  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), function () {
    var lapse = Date.now() - start;
    assert.true((0, _testUtils.nearlyEqual)(lapse, MILLIS_STREAMING_DOWN_OCCUPANCY + settings.scheduler.featuresRefreshRate), 'fetch due to first fallback to polling');
    return {
      status: 200,
      body: {
        splits: [],
        since: 1457552620999,
        till: 1457552620999
      }
    };
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552620999'), {
    status: 200,
    body: {
      since: 1457552620999,
      till: 1457552620999,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // split and segment sync due to streaming up (OCCUPANCY event)

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), {
    status: 200,
    body: {
      splits: [],
      since: 1457552620999,
      till: 1457552620999
    }
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552620999'), {
    status: 200,
    body: {
      since: 1457552620999,
      till: 1457552621999,
      name: 'employees',
      added: ['other_key'],
      removed: []
    }
  }); // extra retry due to double request (greedy fetch until since === till)

  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552621999'), {
    status: 200,
    body: {
      since: 1457552621999,
      till: 1457552621999,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // fetch due to SPLIT_UPDATE event

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), function () {
    var lapse = Date.now() - start;
    assert.true((0, _testUtils.nearlyEqual)(lapse, MILLIS_SPLIT_UPDATE_EVENT_DURING_PUSH), 'sync due to SPLIT_UPDATE event');
    return {
      status: 200,
      body: _splitchangesRealUpdateWithSegments.default
    };
  }); // fetches due to second fallback to polling

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552649999'), {
    status: 200,
    body: {
      splits: [],
      since: 1457552649999,
      till: 1457552649999
    }
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552621999'), {
    status: 200,
    body: {
      since: 1457552621999,
      till: 1457552621999,
      name: 'employees',
      added: [],
      removed: []
    }
  });
  fetchMock.getOnce(settings.url('/splitChanges?since=1457552649999'), function () {
    var lapse = Date.now() - start;
    assert.true((0, _testUtils.nearlyEqual)(lapse, MILLIS_STREAMING_PAUSED_CONTROL + settings.scheduler.featuresRefreshRate), 'fetch due to second fallback to polling');
    return {
      status: 200,
      body: {
        splits: [],
        since: 1457552649999,
        till: 1457552649999
      }
    };
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552621999'), {
    status: 200,
    body: {
      since: 1457552621999,
      till: 1457552621999,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // split and segment sync due to streaming up (CONTROL event)

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552649999'), {
    status: 200,
    body: {
      splits: [],
      since: 1457552649999,
      till: 1457552649999
    }
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552621999'), {
    status: 200,
    body: {
      since: 1457552621999,
      till: 1457552621999,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // fetch due to SEGMENT_UPDATE event

  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552621999'), {
    status: 200,
    body: {
      since: 1457552621999,
      till: 1457552650000,
      name: 'employees',
      added: [],
      removed: [key]
    }
  }); // extra retry (fetch until since === till)

  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552650000'), {
    status: 200,
    body: {
      since: 1457552650000,
      till: 1457552650000,
      name: 'employees',
      added: [],
      removed: []
    }
  }); // fetches due to third fallback to polling

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552649999'), {
    status: 200,
    body: {
      splits: [],
      since: 1457552649999,
      till: 1457552649999
    }
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552650000'), {
    status: 200,
    body: {
      since: 1457552650000,
      till: 1457552650000,
      name: 'employees',
      added: [],
      removed: []
    }
  });
  fetchMock.getOnce(settings.url('/splitChanges?since=1457552649999'), function () {
    var lapse = Date.now() - start;
    assert.true((0, _testUtils.nearlyEqual)(lapse, MILLIS_STREAMING_DISABLED_CONTROL + settings.scheduler.featuresRefreshRate), 'fetch due to third fallback to polling');
    return {
      status: 200,
      body: _splitchangesRealUpdateWithoutSegments.default
    };
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552650000'), {
    status: 200,
    body: {
      since: 1457552650000,
      till: 1457552650000,
      name: 'employees',
      added: [],
      removed: []
    }
  });
  fetchMock.getOnce(settings.url('/splitChanges?since=1457552669999'), function () {
    var lapse = Date.now() - start;
    assert.true((0, _testUtils.nearlyEqual)(lapse, MILLIS_STREAMING_DISABLED_CONTROL + settings.scheduler.featuresRefreshRate * 2), 'fetch due to third fallback to polling');
    return {
      status: 200,
      body: {
        splits: [],
        since: 1457552669999,
        till: 1457552669999
      }
    };
  });
  fetchMock.getOnce(settings.url('/segmentChanges/employees?since=1457552650000'), {
    status: 200,
    body: {
      since: 1457552650000,
      till: 1457552650000,
      name: 'employees',
      added: [],
      removed: []
    }
  });
  (0, _testUtils.mockSegmentChanges)(fetchMock, new RegExp("".concat(settings.url('/segmentChanges'), "/(splitters|developers)")), [key]);
  fetchMock.get(new RegExp('.*'), function (url) {
    assert.fail('unexpected GET request with url: ' + url);
  });
  fetchMock.post('*', 200);
  start = Date.now();
  splitio = (0, _index.SplitFactory)(config);
  client = splitio.client();
}
