// @splitsoftware/splitio/es/__tests__/browserSuites/fetch-specific-splits.spec.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { SplitFactory } from '../../';
import { splitFilters, queryStrings, groupedFilters } from '../mocks/fetchSpecificSplits'; // import splitChangesMock1 from '../mocks/splitchanges.since.-1.json';

var baseConfig = {
  core: {
    authorizationKey: '<fake-token-push-1>',
    key: 'nicolas@split.io'
  },
  scheduler: {
    featuresRefreshRate: 0.01
  },
  streamingEnabled: false
};
export default function fetchSpecificSplits(fetchMock, assert) {
  assert.plan(splitFilters.length);

  var _loop = function _loop(i) {
    var urls = {
      sdk: 'https://sdkurl' + i
    };

    var config = _objectSpread(_objectSpread({}, baseConfig), {}, {
      sync: {
        splitFilters: splitFilters[i]
      },
      urls: urls
    });

    if (groupedFilters[i]) {
      // tests where validateSplitFilters executes normally
      var queryString = queryStrings[i] || '';
      var factory;
      fetchMock.getOnce(urls.sdk + '/splitChanges?since=-1' + queryString, {
        status: 200,
        body: {
          splits: [],
          since: -1,
          till: 1457552620999
        }
      });
      fetchMock.getOnce(urls.sdk + '/splitChanges?since=1457552620999' + queryString, {
        status: 200,
        body: {
          splits: [],
          since: 1457552620999,
          till: 1457552620999
        }
      });
      fetchMock.getOnce(urls.sdk + '/splitChanges?since=1457552620999' + queryString, function () {
        factory.client().destroy().then(function () {
          assert.pass("splitFilters #".concat(i));
        });
        return {
          status: 200,
          body: {
            splits: [],
            since: 1457552620999,
            till: 1457552620999
          }
        };
      });
      fetchMock.get(urls.sdk + '/mySegments/nicolas%40split.io', {
        status: 200,
        body: {
          'mySegments': []
        }
      });
      factory = SplitFactory(config);
    } else {
      // tests where validateSplitFilters throws an exception
      try {
        SplitFactory(config);
      } catch (e) {
        assert.equal(e.message, queryStrings[i]);
      }
    }
  };

  for (var i = 0; i < splitFilters.length; i++) {
    _loop(i);
  }
}
