// @splitsoftware/splitio/lib/__tests__/nodeSuites/impressions.spec.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _ = require("../../");

var _settings = _interopRequireDefault(require("../../utils/settings"));

var _labels = require("../../utils/labels");

var _splitchangesSince = _interopRequireDefault(require("../mocks/splitchanges.since.-1.json"));

var _splitchangesSince2 = _interopRequireDefault(require("../mocks/splitchanges.since.1457552620999.json"));

var _constants = require("../../utils/constants");

var _time = require("../../utils/time");

var baseUrls = {
  sdk: 'https://sdk.baseurl/impressionsSuite',
  events: 'https://events.baseurl/impressionsSuite'
};
var settings = (0, _settings.default)({
  core: {
    key: '<fake id>'
  },
  urls: baseUrls,
  streamingEnabled: false
});
var config = {
  core: {
    authorizationKey: '<fake-token-2>'
  },
  scheduler: {
    featuresRefreshRate: 1,
    segmentsRefreshRate: 1,
    metricsRefreshRate: 3000,
    impressionsRefreshRate: 5
  },
  urls: baseUrls,
  startup: {
    eventsFirstPushWindow: 3000
  },
  streamingEnabled: false
};
var truncatedTimeFrame;

function _default(_x, _x2, _x3) {
  return _ref.apply(this, arguments);
}

function _ref() {
  _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(key, fetchMock, assert) {
    var splitio, client, evaluationsStart, readyEvaluationsStart, evaluationsEnd;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            // Mocking this specific route to make sure we only get the items we want to test from the handlers.
            fetchMock.getOnce(settings.url('/splitChanges?since=-1'), {
              status: 200,
              body: _splitchangesSince.default
            });
            fetchMock.get(settings.url('/splitChanges?since=1457552620999'), {
              status: 200,
              body: _splitchangesSince2.default
            });
            fetchMock.get(new RegExp("".concat(settings.url('/segmentChanges/'), ".*")), {
              status: 200,
              body: {
                since: 10,
                till: 10,
                name: 'segmentName',
                added: [],
                removed: []
              }
            });
            splitio = (0, _.SplitFactory)(config);
            client = splitio.client();
            evaluationsStart = 0, readyEvaluationsStart = 0, evaluationsEnd = 0;
            fetchMock.postOnce(settings.url('/testImpressions/bulk'), function (url, opts) {
              assert.equal(opts.headers.SplitSDKImpressionsMode, _constants.OPTIMIZED);
              var data = JSON.parse(opts.body);
              assert.equal(data.length, 3, 'We performed evaluations for three splits, so we should have 3 items total.'); // finding these validate the feature names collection too

              var dependencyChildImpr = data.filter(function (e) {
                return e.f === 'hierarchical_splits_test';
              })[0];
              var alwaysOnWithConfigImpr = data.filter(function (e) {
                return e.f === 'split_with_config';
              })[0];
              var notExistentSplitImpr = data.filter(function (e) {
                return e.f === 'not_existent_split';
              })[0];
              assert.equal(notExistentSplitImpr.i.length, 1); // Only one, the split not found is filtered by the non existent Split check.

              assert.equal(alwaysOnWithConfigImpr.i.length, 2);
              assert.equal(dependencyChildImpr.i.length, 1);
              assert.true(dependencyChildImpr, 'Split we wanted to evaluate should be present on the impressions.');
              assert.false(data.some(function (e) {
                return e.f === 'hierarchical_dep_always_on';
              }), 'Parent split evaluations should not result in impressions.');
              assert.false(data.some(function (e) {
                return e.f === 'hierarchical_dep_hierarchical';
              }), 'No matter how deep is the chain.');
              assert.true(alwaysOnWithConfigImpr, 'Split evaluated with config should have generated an impression too.');
              assert.false(Object.prototype.hasOwnProperty.call(alwaysOnWithConfigImpr.i[0], 'configuration'), 'Impressions do not change with configuration evaluations.');
              assert.false(Object.prototype.hasOwnProperty.call(alwaysOnWithConfigImpr.i[0], 'config'), 'Impressions do not change with configuration evaluations.');

              function validateImpressionData(output, expected) {
                var performedWhenReady = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
                assert.equal(output.k, expected.keyName, 'Present impressions should have the correct key.');
                assert.equal(output.b, expected.bucketingKey, 'Present impressions should have the correct bucketingKey.');
                assert.equal(output.t, expected.treatment, 'Present impressions should have the correct treatment.');
                assert.equal(output.r, expected.label, 'Present impressions should have the correct label.');
                assert.equal(output.c, expected.changeNumber, 'Present impressions should have the correct changeNumber.');
                assert.true(output.m >= (performedWhenReady ? readyEvaluationsStart : evaluationsStart) && output.m <= evaluationsEnd, 'Present impressions should have the correct timestamp (test with error margin).');
              }

              validateImpressionData(notExistentSplitImpr.i[0], {
                keyName: 'facundo@split.io',
                label: _labels.SDK_NOT_READY,
                treatment: 'control',
                bucketingKey: undefined,
                changeNumber: undefined
              }, false);
              validateImpressionData(dependencyChildImpr.i[0], {
                keyName: 'facundo@split.io',
                label: 'expected label',
                treatment: 'on',
                bucketingKey: undefined,
                changeNumber: 2828282828
              });
              validateImpressionData(alwaysOnWithConfigImpr.i[0], {
                keyName: 'facundo@split.io',
                label: 'another expected label',
                treatment: 'on',
                bucketingKey: 'test_buck_key',
                changeNumber: 828282828282
              }); // Not push impressions with a invalid key (aka matching key)

              assert.true(dependencyChildImpr.i.filter(function (e) {
                return e.k !== 'facundo@split.io';
              }).length === 0, 'There should be impressions with valid keys, the sdk will not push a impression with invalid a key');
              client.destroy();
              assert.end();
              return 200;
            });
            fetchMock.postOnce(settings.url('/testImpressions/bulk'), 200);
            fetchMock.postOnce(settings.url('/testImpressions/count'), function (url, opts) {
              var data = JSON.parse(opts.body);
              assert.equal(data.pf.length, 3, 'We should generated impressions for three features.'); // finding these validate the feature names collection too

              var dependencyChildImpr = data.pf.filter(function (e) {
                return e.f === 'hierarchical_splits_test';
              })[0];
              var alwaysOnWithConfigImpr = data.pf.filter(function (e) {
                return e.f === 'split_with_config';
              })[0];
              var notExistentSplitImpr = data.pf.filter(function (e) {
                return e.f === 'not_existent_split';
              })[0];
              assert.equal(dependencyChildImpr.rc, 1);
              assert.equal((0, _typeof2.default)(dependencyChildImpr.m), 'number');
              assert.equal(dependencyChildImpr.m, truncatedTimeFrame);
              assert.equal(alwaysOnWithConfigImpr.rc, 3);
              assert.equal((0, _typeof2.default)(alwaysOnWithConfigImpr.m), 'number');
              assert.equal(alwaysOnWithConfigImpr.m, truncatedTimeFrame);
              assert.equal(notExistentSplitImpr.rc, 1);
              assert.equal((0, _typeof2.default)(notExistentSplitImpr.m), 'number');
              assert.equal(notExistentSplitImpr.m, truncatedTimeFrame);
              return 200;
            });
            fetchMock.postOnce(settings.url('/testImpressions/count'), 200);
            splitio.Logger.enable();
            evaluationsStart = Date.now();
            assert.equal(client.getTreatment(key, 'not_existent_split'), 'control', "If we try to get an evaluation BEFORE the client is ready, we expect ".concat(_labels.SDK_NOT_READY, " label on the impression."));
            _context.next = 15;
            return client.ready();

          case 15:
            readyEvaluationsStart = Date.now();
            truncatedTimeFrame = (0, _time.truncateTimeFrame)(readyEvaluationsStart); // depends on hierarchical_dep_hierarchical which depends on hierarchical_dep_always_on

            assert.equal(client.getTreatment(key, 'not_existent_split'), 'control', 'If we try to get an evaluation for a non existent split AFTER the client is ready, it won\'t log an impression.');
            assert.equal(client.getTreatment(key, 'hierarchical_splits_test'), 'on', 'We should get an evaluation as always.');
            assert.equal(client.getTreatment(false, 'hierarchical_splits_test'), 'control', 'We should return control with a invalid key and that impression is not tracked.');
            assert.deepEqual(client.getTreatmentWithConfig({
              matchingKey: key,
              bucketingKey: 'test_buck_key'
            }, 'split_with_config'), {
              // I'll run this one with bucketing key.
              treatment: 'on',
              config: '{"color":"brown","dimensions":{"height":12,"width":14},"text":{"inner":"click me"}}'
            }, 'We should get an evaluation as always.');
            client.getTreatmentWithConfig({
              matchingKey: key,
              bucketingKey: 'test_buck_key'
            }, 'split_with_config');
            client.getTreatmentWithConfig({
              matchingKey: 'different',
              bucketingKey: 'test_buck_key'
            }, 'split_with_config');
            splitio.Logger.disable();
            evaluationsEnd = Date.now();

          case 25:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _ref.apply(this, arguments);
}
