// @splitsoftware/splitio/es/sync/__tests__/mocks/eventSourceMock.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";

/**
 * EventEmitter mock based on https://github.com/gcedo/eventsourcemock/blob/master/src/EventSource.js
 *
 * To setup the mock assign it to the window object.
 * ```
 *  import EventSource from 'eventsourcemock';
 *  Object.defineProperty(window, 'EventSource', {
 *    value: EventSource,
 *  });
 * ```
 *
 */
import EventEmitter from 'events';
var defaultOptions = {
  withCredentials: false
};
export var sources = {};

var __listener;

export function setMockListener(listener) {
  __listener = listener;
}

var EventSource = /*#__PURE__*/function () {
  function EventSource(url) {
    var configuration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultOptions;

    _classCallCheck(this, EventSource);

    this.url = url;
    this.withCredentials = configuration.withCredentials;
    this.readyState = 0; // eslint-disable-next-line no-undef

    this.__emitter = new EventEmitter();
    sources[url] = this;
    if (__listener) setTimeout(__listener, 0, this);
  }

  _createClass(EventSource, [{
    key: "addEventListener",
    value: function addEventListener(eventName, listener) {
      this.__emitter.addListener(eventName, listener);
    }
  }, {
    key: "removeEventListener",
    value: function removeEventListener(eventName, listener) {
      this.__emitter.removeListener(eventName, listener);
    }
  }, {
    key: "close",
    value: function close() {
      this.readyState = 2;
    } // The following methods can be used to mock EventSource behavior and events

  }, {
    key: "emit",
    value: function emit(eventName, messageEvent) {
      this.__emitter.emit(eventName, messageEvent);
    }
  }, {
    key: "emitError",
    value: function emitError(error) {
      if (typeof this.onerror === 'function') {
        this.onerror(error);
      }
    }
  }, {
    key: "emitOpen",
    value: function emitOpen() {
      this.readyState = 1;

      if (typeof this.onopen === 'function') {
        this.onopen();
      }
    }
  }, {
    key: "emitMessage",
    value: function emitMessage(message) {
      if (typeof this.onmessage === 'function') {
        this.onmessage(message);
      }
    }
  }]);

  return EventSource;
}();

export { EventSource as default };
EventSource.CONNECTING = 0;
EventSource.OPEN = 1;
EventSource.CLOSED = 2;
